libs/@local/harpc/client/typescript/src/types/ErrorCode.ts

```diff
@@ -0,0 +1,109 @@
+import type { FastCheck } from "effect";
+import {
+  Data,
+  Effect,
+  Equal,
+  Hash,
+  Inspectable,
+  pipe,
+  Pipeable,
+  Predicate,
+} from "effect";
+
+import { U16_MAX } from "../constants.js";
+import { createProto, encodeDual } from "../utils.js";
+import * as Buffer from "../wire-protocol/Buffer.js";
+
+const TypeId = Symbol("@local/harpc-client/wire-protocol/types/ErrorCode");
+export type TypeId = typeof TypeId;
+
+export class ErrorCodeTooLarge extends Data.TaggedError("ErrorCodeTooLarge")<{
+  received: number;
+}> {
+  get message() {
+    return `error code is too large, expected error code to be less than or equal to ${U16_MAX}, got ${this.received}`;
+  }
+}
+
+export class ErrorCodeTooSmall extends Data.TaggedError(
+  "ErrorCodeNotPositive",
+)<{ received: number }> {
+  get message() {
+    return `error code mut be a positive number, got ${this.received}`;
+  }
+}
+
+export interface ErrorCode
+  extends Equal.Equal,
+    Inspectable.Inspectable,
+    Pipeable.Pipeable {
+  readonly [TypeId]: TypeId;
+
+  readonly code: number;
+}
+
+const ErrorCodeProto: Omit<ErrorCode, "code"> = {
+  [TypeId]: TypeId,
+
+  [Equal.symbol](this: ErrorCode, that: Equal.Equal) {
+    // eslint-disable-next-line @typescript-eslint/no-use-before-define
+    return isErrorCode(that) && Equal.equals(this.code, that.code);
+  },
+
+  [Hash.symbol](this: ErrorCode) {
+    return pipe(
+      Hash.hash(this[TypeId]),
+      Hash.combine(Hash.number(this.code)),
+      Hash.cached(this),
```

Comment:
we're implementing `Hash.Hash` on a type because `Equal.Equal` extends from `Hash.Hash`, see:

* https://effect.website/docs/trait/equal/
* https://effect.website/docs/trait/hash/
* https://effect-ts.github.io/effect/effect/Hash.ts.html#cached

Specifically I made use of `Hash.cached` here, because the very similar [`Duration`](https://github.com/Effect-TS/effect/blob/d8356aad428a0c2290db52380220f81d9ec94232/packages/effect/src/Duration.ts#L144-L146) type and _many many_ other similar types in effect make use of it.